core/src/impl/Kokkos_GraphNodeImpl.hpp

```diff
@@ -0,0 +1,296 @@
+
+// Specialization for the case with the concrete type of the kernel, but the
+// predecessor erased.
+template <class ExecutionSpace, class Kernel>
+struct GraphNodeImpl<ExecutionSpace, Kernel,
+                     Kokkos::Experimental::TypeErasedTag>
+    : GraphNodeImpl<ExecutionSpace, Kokkos::Experimental::TypeErasedTag,
+                    Kokkos::Experimental::TypeErasedTag> {
+ private:
+  using base_t =
+      GraphNodeImpl<ExecutionSpace, Kokkos::Experimental::TypeErasedTag,
+                    Kokkos::Experimental::TypeErasedTag>;
+
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public member types"> {{{2
+
+  using node_ref_t =
+      Kokkos::Experimental::GraphNodeRef<ExecutionSpace, Kernel,
+                                         Kokkos::Experimental::TypeErasedTag>;
+  using kernel_type = Kernel;
+
+  // </editor-fold> end public member types }}}2
+  //----------------------------------------------------------------------------
+
+ private:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private data members"> {{{2
+
+  Kernel m_kernel;
+
+  // </editor-fold> end private data members }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="Ctors, destructors, and assignment"> {{{2
+
+  template <class KernelDeduced>
+  GraphNodeImpl(ExecutionSpace const& ex, _graph_node_kernel_ctor_tag,
+                KernelDeduced&& arg_kernel)
+      : base_t(ex), m_kernel((KernelDeduced &&) arg_kernel) {}
+
+  template <class... Args>
+  GraphNodeImpl(ExecutionSpace const& ex, _graph_node_is_root_ctor_tag,
+                Args&&... args)
+      : base_t(ex, _graph_node_is_root_ctor_tag{}, (Args &&) args...) {}
+
+  //- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  // <editor-fold desc="Rule of 6 for not copyable or movable"> {{{3
+
+  // Not copyable or movable
+  GraphNodeImpl()                         = delete;
+  GraphNodeImpl(GraphNodeImpl const&)     = delete;
+  GraphNodeImpl(GraphNodeImpl&&) noexcept = delete;
+  GraphNodeImpl& operator=(GraphNodeImpl const&) = delete;
+  GraphNodeImpl& operator=(GraphNodeImpl&&) noexcept = delete;
+  ~GraphNodeImpl() override                          = default;
+
+  // </editor-fold> end Rule of 6 for not copyable or movable }}}3
+  //- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+  // </editor-fold> end Ctors, destructors, and assignment }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="member accessors"> {{{2
+
+  Kernel& get_kernel() & { return m_kernel; }
+  Kernel const& get_kernel() const& { return m_kernel; }
```

Comment:
Please explain why these need to be ref-qualified

Reply:
They return lvalue references to something that's part of the object representation, so an rvalue overload would be a dangling reference. Would you like me to add a comment to this effect?

Reply:
```suggestion
  // Reference qualified to prevent dangling reference to data member
  Kernel& get_kernel() & { return m_kernel; }
  Kernel const& get_kernel() const& { return m_kernel; }
```

Reply:
```suggestion
  // Reference qualified to prevent dangling reference to data member
  Kernel& get_kernel() & { return m_kernel; }
  Kernel const& get_kernel() const& { return m_kernel; }
  Kernel&& get_kernel() && = delete;
```

Reply:
Resolved in 75de11b